src/replica/HttpModuleBase.cc

```diff
@@ -63,10 +64,51 @@ void HttpModuleBase::execute(string const& subModuleName, HttpAuthType const aut
     }
 }
 
+void HttpModuleBase::checkApiVersion(string const& func, unsigned int minVersion) const {
+    unsigned int const maxVersion = HttpMetaModule::version;
+    unsigned int version = 0;
+    string const versionAttrName = "version";
+    json const errorEx = json::object({{"min_version", minVersion}, {"max_version", maxVersion}});
+
+    // Intercept exceptions thrown when converting the attribute's value (if provided)
+    // in order to inject the allowed range of the version numbers into the extended
+    // error sent back to the caller.
+    //
+    // Note that requests sent w/o explicitly specified API version will still be
+    // processed. In this case a warning will be sent in the response object.
+    try {
+        if (req()->method == "GET") {
+            if (!query().has(versionAttrName)) {
+                _warningOnVersionMissing = "No version number was provided in the request's query.";
+                warn(_warningOnVersionMissing);
+                return;
+            }
+            version = query().requiredUInt(versionAttrName);
+        } else {
+            if (!body().has(versionAttrName)) {
+                _warningOnVersionMissing = "No version number was provided in the request's body.";
+                warn(_warningOnVersionMissing);
+                return;
+            }
+            version = body().required<unsigned int>(versionAttrName);
+        }
+    } catch (...) {
+        throw HttpError(func, "The required parameter " + versionAttrName + " is not a number.", errorEx);
+    }
+    if (!(minVersion <= version && version <= maxVersion)) {
+        throw HttpError(func,
+                        "The requested version " + to_string(version) +
+                                " of the API is not in the range supported by the service.",
```

Comment:
Maybe return the supported range in the error message here too?

Reply:
Actually, the code already does this. In the expanded version of the method you would see:
```
json const errorEx = json::object({{"min_version", minVersion}, {"max_version", maxVersion}});
try {
    ...
} catch (...) {
    throw HttpError(func, "The required parameter " + versionAttrName + " is not a number.", errorEx);
}
if (!(minVersion <= version && version <= maxVersion)) {
    throw HttpError(func,
                    "The requested version " + to_string(version) +
                            " of the API is not in the range supported by the service.",
                    errorEx);
}
```
The exception class has the third parameter that accepts an optional dictionary that depends on an error.